File: packages/core/src/features/vector.ts
import type { DigraphStats } from "../types/features";
import { mean, stdDev, median, percentile } from "./utils";

export function digraphStatsFromSamples(
  samples: number[],
): DigraphStats | null {
  if (samples.length < 5) return null;

  const sorted = [...samples].sort((a, b) => a - b);
  const n = sorted.length;
  const m = sorted.reduce((sum, v) => sum + v, 0) / n;
  const variance = sorted.reduce((sum, v) => sum + (v - m) ** 2, 0) / (n - 1);

  return {
    count: n,
    mean: m,
    std_dev: Math.sqrt(variance),
    median: sorted[Math.floor(n / 2)],
    p10: sorted[Math.floor(n * 0.1)],
    p90: sorted[Math.min(Math.floor(n * 0.9), n - 1)],
  };
}

export { mean, stdDev, median, percentile };

File: apps/desktop/src-tauri/src/features/vector.rs
use serde::{Deserialize, Serialize};

use super::tier1::Tier1Features;
use super::tier2::Tier2Features;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FeatureVector {
    pub version: String,
    pub window_start_ms: f64,
    pub window_end_ms: f64,
    pub keystroke_count: u32,
    pub tier1: Tier1Features,
    pub tier2: Tier2Features,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DigraphStats {
    pub count: u32,
    pub mean: f64,
    pub std_dev: f64,
    pub median: f64,
    pub p10: f64,
    pub p90: f64,
}

impl DigraphStats {
    pub fn from_samples(samples: &mut Vec<f64>) -> Option<Self> {
        if samples.len() < 5 {
            return None;
        }
        samples.sort_by(|a, b| a.partial_cmp(b).unwrap_or(std::cmp::Ordering::Equal));

        let n = samples.len() as f64;
        let mean = samples.iter().sum::<f64>() / n;
        let variance = samples.iter().map(|x| (x - mean).powi(2)).sum::<f64>() / (n - 1.0);
        let std_dev = variance.sqrt();
        let median = samples[samples.len() / 2];
        let p10 = samples[(samples.len() as f64 * 0.1) as usize];
        let p90 = samples[(samples.len() as f64 * 0.9).min(samples.len() as f64 - 1.0) as usize];

        Some(DigraphStats {
            count: samples.len() as u32,
            mean,
            std_dev,
            median,
            p10,
            p90,
        })
    }
}
